packages/trpc/server/routers/viewer/teams/removeMember/PBACRemoveMemberService.ts

```diff
@@ -0,0 +1,60 @@
+import { isTeamOwner } from "@calcom/features/ee/teams/lib/queries";
+import { PermissionMapper } from "@calcom/features/pbac/domain/mappers/PermissionMapper";
+import { Resource, CustomAction } from "@calcom/features/pbac/domain/types/permission-registry";
+import { PermissionCheckService } from "@calcom/features/pbac/services/permission-check.service";
+
+import { TRPCError } from "@trpc/server";
+
+import { BaseRemoveMemberService } from "./BaseRemoveMemberService";
+import type { RemoveMemberContext, RemoveMemberPermissionResult } from "./IRemoveMemberService";
+
+export class PBACRemoveMemberService extends BaseRemoveMemberService {
+  private permissionService = new PermissionCheckService();
+
+  async checkRemovePermissions(context: RemoveMemberContext): Promise<RemoveMemberPermissionResult> {
+    const { userId, teamIds, isOrg } = context;
+
+    const resource = isOrg ? Resource.Organization : Resource.Team;
+    const removePermission = PermissionMapper.toPermissionString({
+      resource,
+      action: CustomAction.Remove,
+    });
+
+    const teamsWithPermission = await this.permissionService.getTeamIdsWithPermission(
+      userId,
+      removePermission
+    );
```

Comment:
This is now O(1) instead of O(n+1) as we are not calling checkPermissions in a loop